### PR TITLE
Introduce a "dynamic" block size limit for proposing

### PR DESCRIPTION
**Pull request by bkchr**
This adds support for using a dynamic block size limit per call to
`propose`. This is required for Cumulus/Parachains to always use stay in
the limits of the maximum allowed PoV size.

As described in the docs, the block limit is only checked in the process
of pushing transactions. As we normally do some other operations in
`on_finalize`, it can happen that the block size still grows when there
is some proof being collected (as we do for parachains). This means,
that the given block limit needs to be rather conservative on the actual
value and should not be the upper limit.

polkadot companion: https://github.com/paritytech/polkadot/pull/2883

### CODE REVIEW

File: client/basic-authorship/src/basic_authorship.rs [hunk 1]
```diff
@@ -141,7 +158,7 @@ impl<B, Block, C, A, PR> ProposerFactory<A, B, C, PR>
 
 		let id = BlockId::hash(parent_hash);
 
-		info!("🙌 Starting consensus session on top of parent {:?}", parent_hash);
+		info!("🙌 Starting consdefault_block_size_limiton top of parent {:?}", parent_hash);
```
Comment (ordian) [inline client/basic-authorship/src/basic_authorship.rs:161]:
```suggestion
		info!("🙌 Starting consensus session on top of parent {:?}", parent_hash);
```

File: primitives/state-machine/src/proving_backend.rs [hunk 2]
```diff
@@ -109,9 +109,55 @@ impl<'a, S, H> ProvingBackendRecorder<'a, S, H>
...
+	pub fn get(&self, key: &Hash) -> Option<Option<DBValue>> {
+		self.records.read().get(key).cloned()
+	}
+
+	/// Returns the estimated encoded size of the proof.
+	///
+	/// The estimation is maybe bigger (by in maximum 4 bytes), but never smaller than the actual
+	/// encoded proof.
+	pub fn estimate_encoded_size(&self) -> usize {
+		self.encoded_size.load(Ordering::Relaxed)
```
Comment (ordian) [inline primitives/state-machine/src/proving_backend.rs:141]: why `Ordering::Relaxed` everywhere?

Comment (bkchr) [reply primitives/state-machine/src/proving_backend.rs:141]: Because we don't need any other ordering?

**Review by kianenigma — COMMENTED** [review]
I breifly skimmed this and it does not seem to that the runtime is telling how big block should be. Why? 

It was always strange to me that the propser blindly tries to add transactions until some opque block limit is reached (weight and length). I agree that block weight should not be exposed to the client directly as it is frame dependent. But weight is pretty generic and unopinionated. 

But seeing the title of the PR, my assumption was that there would be a new api here through which the client would ask the runtime how big the block would be. 

What is the rational here?

File: primitives/state-machine/src/proving_backend.rs (hunk 2 above)
Comment (ordian) [reply primitives/state-machine/src/proving_backend.rs:143]: I just don't get why everything in substrate is using relaxed ordering.
You can also put encoded_size under `Arch<RwLock<>>` along with HashMap, so that it doesn't need to be an atomic that goes out of sync if you call `record` and then `estimate_encoded_size`

**Comment by bkchr** [comment]
> I breifly skimmed this and it does not seem to that the runtime is telling how big block should be. Why?

The runtime can still dictate this. However we already had some client limit before (4MiB) and than there is also the next limit 16MiB from the networking. In the end the client will always have control over this and the runtime can not do that much against it.

> It was always strange to me that the propser blindly tries to add transactions until some opque block limit is reached (weight and length). I agree that block weight should not be exposed to the client directly as it is frame dependent. But weight is pretty generic and unopinionated.

But what if your runtime doesn't use any weight? The current proposer design is also rather dumb, but it works currently without any problems ;) If we encounter some problems, we can still make it smarterl

> 
> But seeing the title of the PR, my assumption was that there would be a new api here through which the client would ask the runtime how big the block would be.
> 
> What is the rational here?

This pr is mainly for parachains. For parachains we have the PoV size limit that is dictated by the relay chain and the PoV not only contains the header & the extrinsics, it also contains the proof. 

To give a little bit more context. I fought a little bit with myself about this and I'm also far from being really happy with the solution, but it is also the solution that will not require any crazy sort of migrations. So, I also thought of introducing some host function that enables the runtime to ask what the current proof size is and let the block production stop. The problem with that is that we need a new host function and if we decide to change something in regards to this in the future, we will still need to stick with this function until the end of the universe, aka not such a good option :D Next point is that the client/node here is again also interested into staying in these bounds, because the relay chain otherwise will just not accept its PoV and the collator will not get payed etc. So, there is an incentivize that the collator sticks to the maximum size. 

Currently it is also relative dumb and we just hope to not break any size limits, however this will improve with the new benchmarking that will also take the proof size into account. If we have all the benchmarking updated to also track the proof size, we will have a much better weight and can much better predict from a runtime if we should accept a certain transaction etc.

TLDR: This is just the beginning, which will enable us to switch relative easy to much more sophisticated solutions in the future.

**Comment by kianenigma** [comment]
> But what if your runtime doesn't use any weight? The current proposer design is also rather dumb, but it works currently without any problems ;) If we encounter some problems, we can still make it smarterl

I agree, in my sentence: 

> I agree that block weight should not be exposed to the client directly as it is frame dependent. But weight is pretty generic and unopinionated.

what I really wanted to say was

> I agree that block weight should not be exposed to the client directly as it is frame dependent. But **length** is pretty generic and unopinionated.

File: primitives/state-machine/src/proving_backend.rs (hunk 2 above)
Comment (bkchr) [reply primitives/state-machine/src/proving_backend.rs:143]: > I just don't get why everything in substrate is using relaxed ordering.

If there is no requirement that an atomic needs to be executed in some special order, we can just `Relaxed`. Here we clearly don't need any special ordering, because nothing is dependent on being updated before some mutex is locked/unlocked.

> You can also put encoded_size under Arch<RwLock<>> along with HashMap, so that it doesn't need to be an atomic that goes out of sync if you call record and then estimate_encoded_size

I know, but as you see here, I decided against this. I first thought of creating an `ProofRecorderInner` that I would put behind the mutex, but I don't wanted to create another struct etc. So, I decided to use an `Atomic`. 

The whole proof recorder could also just be some `Rc`, as we always execute block production single threaded. This stuff here is only Send/Sync for futures.

Comment (bkchr) [reply primitives/state-machine/src/proving_backend.rs]: I removed the atomic.

File: client/block-builder/src/lib.rs [hunk 3]
```diff
@@ -135,6 +135,8 @@ pub struct BlockBuilder<'a, Block: BlockT, A: ProvideRuntimeApi<Block>, B> {
 	block_id: BlockId<Block>,
 	parent_hash: Block::Hash,
 	backend: &'a B,
+	/// The estimated size of the block header.
+	estimated_header_size: usize,
```
Comment (andresilva) [inline client/block-builder/src/lib.rs:139]: I am OK with this but it would probably be fine if we just computed it anyway anytime it is required in `estimate_block_size`.

File: client/basic-authorship/src/basic_authorship.rs [hunk 4]
```diff
@@ -311,8 +335,13 @@ impl<A, B, Block, C, PR> Proposer<B, Block, C, A, PR>
 			},
 		};
 
+		let block_size_limit = block_size_limit.unwrap_or(self.default_block_size_limit);
+
 		debug!("Attempting to push transactions from the pool.");
 		debug!("Pool status: {:?}", self.transaction_pool.status());
+		let mut transaction_pushed = true;
```
Comment (andresilva) [inline client/basic-authorship/src/basic_authorship.rs:342]:
```suggestion
		let mut transaction_pushed = false;
```

File: client/basic-authorship/src/basic_authorship.rs [hunk 5]
```diff
@@ -324,9 +353,30 @@ impl<A, B, Block, C, PR> Proposer<B, Block, C, A, PR>
...
+				}
+			}
+
 			trace!("[{:?}] Pushing to the block.", pending_tx_hash);
 			match sc_block_builder::BlockBuilder::push(&mut block_builder, pending_tx_data) {
 				Ok(()) => {
+					transaction_pushed = true;
 					debug!("[{:?}] Pushed to the block.", pending_tx_hash);
 				}
 				Err(ApplyExtrinsicFailed(Validity(e)))
```
Comment (andresilva) [inline client/basic-authorship/src/basic_authorship.rs:382]: Does it still make sense to handle it like this?

File: client/consensus/slots/src/lib.rs [hunk 6]
```diff
@@ -313,6 +313,7 @@ pub trait SimpleSlotWorker<B: BlockT> {
 				logs,
 			},
 			proposing_remaining_duration.mul_f32(0.98),
+            None,
```
Comment (andresilva) [inline client/consensus/slots/src/lib.rs:316]:
```suggestion
			None,
```

File: client/consensus/slots/src/slots.rs [hunk 7]
```diff
@@ -58,6 +58,10 @@ pub struct SlotInfo {
 	pub inherent_data: InherentData,
 	/// Slot duration.
 	pub duration: Duration,
+	/// Some potatenial block size limit for the block of this slot.
```
Comment (andresilva) [inline client/consensus/slots/src/slots.rs:61]:
```suggestion
	/// Some potential block size limit for the block to be authored at this slot.
```

File: primitives/consensus/common/src/evaluation.rs [hunk 8]
```diff
@@ -50,17 +47,12 @@ pub fn evaluate_initial<Block: BlockT>(
...
 	parent_hash: &<Block as BlockT>::Hash,
 	parent_number: <<Block as BlockT>::Header as HeaderT>::Number,
-	max_block_size: usize,
 ) -> Result<()> {
 
 	let encoded = Encode::encode(proposal);
 	let proposal = Block::decode(&mut &encoded[..])
 		.map_err(|e| Error::BadProposalFormat(e))?;
 
-	if encoded.len() > max_block_size {
```
Comment (andresilva) [inline primitives/consensus/common/src/evaluation.rs:60]: I guess we could still keep this check around, in case someone implements their own proposer.

**Review by andresilva — CHANGES_REQUESTED** [review]

File: client/basic-authorship/src/basic_authorship.rs [hunk 9]
```diff
@@ -93,6 +100,8 @@ impl<A, B, C> ProposerFactory<A, B, C, EnableProofRecording> {
 	/// Create a new proposer factory with proof recording enabled.
 	///
 	/// Each proposer created by this instance will record a proof while building a block.
+	///
+	/// This will also include the proof into the estimation of the block size. This can be
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:104]: Good question :P

File: client/basic-authorship/src/basic_authorship.rs [hunk 10]
```diff
@@ -264,6 +285,7 @@ impl<A, B, Block, C, PR> Proposer<B, Block, C, A, PR>
 		inherent_data: InherentData,
 		inherent_digests: DigestFor<Block>,
 		deadline: time::Instant,
+		block_size_limit: Option<usize>,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:289]: Yes we have one :D I actually would liked to have done this at initialization of the `ProposerFactory`. However, with parachains the limit is given by the relay chain and this limit is set in the runtime. So, this could change in any relay chain block (actually in a new session, but yeah it is dynamic)

File: client/block-builder/src/lib.rs (hunk 3 above)
Comment (bkchr) [reply client/block-builder/src/lib.rs:139]: PERFORMANCE OPTIMIZATIONS!

(And we also don't store the header)

File: client/basic-authorship/src/basic_authorship.rs (hunk 5 above)
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:383]: Why not? Do you see any problems?

File: primitives/consensus/common/src/evaluation.rs (hunk 8 above)
Comment (bkchr) [reply primitives/consensus/common/src/evaluation.rs:60]: Not sure. In the end they could just do it themself xD

This code should probably deleted anyway at some point.

File: client/block-builder/src/lib.rs (hunk 3 above)
Comment (andresilva) [reply client/block-builder/src/lib.rs:139]: Ah yeah, I missed that part 😅

File: client/basic-authorship/src/basic_authorship.rs (hunk 5 above)
Comment (andresilva) [reply client/basic-authorship/src/basic_authorship.rs:383]: Ah, I was dumb. This one is checking for block weight not being exceeded.

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
Should be good to go now.

**Comment by bkchr** [comment]
bot merge
